serde_derive/src/de.rs

```diff
@@ -1750,10 +1749,19 @@ fn deserialize_untagged_enum_after(
     );
     let fallthrough_msg = cattrs.expecting().unwrap_or(&fallthrough_msg);
 
+    // This may be infallible so we need to provide the error type.
+    let first_attempt = first_attempt.map(|expr| {
+        quote! {
+            if let _serde::__private::Result::<_, __D::Error>::Ok(__ok) = #expr {
+                return _serde::__private::Ok(__ok);
+            }
+        }
+    });
     quote_block! {
         let __content = <_serde::__private::de::Content as _serde::Deserialize>::deserialize(__deserializer)?;
         let __deserializer = _serde::__private::de::ContentRefDeserializer::<__D::Error>::new(&__content);
 
+        #first_attempt
         #(
             if let _serde::__private::Ok(__ok) = #attempts {
```

Comment:
if the type annotation is added here, we can keep sharing the logic between `first_attempt` and `attempts`

Reply:
Indeed! As I noted, I was trying to minimize code change for code paths unrelated to the new untagged variant semantics, but I'm very happy to add the type annotation in all cases if that's preferable. Please let me know.